### ISSUE TITLE
AMP 导致 loss 为 nan

### ISSUE THREAD

**Issue by Muyun99** [FlyEgle/imageclassification#1]
你好，感谢分享，非常有意义的实验

我在将 AMP 加入到我自己的代码中时遇到一个问题：在训练一段时间后其 loss 会突变为 nan，我尝试了调小学习率和 调大batch_size ，但是并没有能够避免这个问题，我想请教一下您这边有没有什么经验

期待您的回复！

实验名称 | 学习率 | 验证集上最优精度 | 训练时间 | 显存占用 | 备注
-- | -- | -- | -- | -- | --
resnet_fp16_lr_0.1 | 0.1 | 0.9553 | 13m27s | 2300MiB | 在 6k iteration 之后 loss 变为 nan
resnet_fp16_lr_0.01 | 0.01 | 0.9455 | 12m59s | 2300MiB | 在 5k iteration 之后 loss 变为 nan
resnet_fp16_lr_0.001 | 0.001 | 0.7126 | 13m31s | 2300MiB | 在 4k iteration 之后 loss 变为 nan
resnet_nofp16_lr_0.1 | 0.1 | 0.9651 | 19m19s |   | 未使用 fp16
resnet_fp16_lr_0.1_v3 | 0.1 | 0.9635 | 12m10s | 4360MiB | batch_size 改为 256，并且 loss 并未变为 nan
resnet_fp16_lr_0.1_v4 | 0.1 | 0.9624 | 50m29s | 4360MiB | batch_size 改为 256，并且将epoch 改为 200，在 8k iteration 之后 loss 变为 nan

**Comment by FlyEgle** [comment]
1、如果你用的是SGD，以0.1和256的bs作为匹配，训练的bs增大，LR也对应增大，如果不增大，对于训练也不会有太大的影响。
2、AMP使用，需要有scalar对loss进行缩放处理。
3、如果大的bs，最好进行warmup训练。
4、排查数据问题，你可以先用我的代码跑一遍，要是没有问题，可以对比一下代码哪里有遗漏

**Comment by Muyun99** [comment]
感谢您的回复，已经破案了，发现是 `optimizer` 重复step了，是我的问题

```python
if cfg.fp16 is True:
    with autocast():
        logits = model(images)
        loss = criterion(logits, labels)
    scaler.scale(loss).backward()
    scaler.step(optimizer)
    scaler.update()
else:
    logits = model(images)
    loss = criterion(logits, labels)
    loss.backward()
optimizer.step()
```

将 `optimizer.step()` 写到 `else` 语句里面即可，再次感谢您的认真回复，很有价值
